### DISCUSSION TITLE
[Messenger / Redis] "Skip" a message (let another worker process it)?

### DISCUSSION THREAD

**Discussion by bpolaszek** [symfony/symfony#48078] [Q&A]
Hello there,

Does anybody know how to _manually_ acknowledge / not acknowledge messages using the Redis transport?

Use case: 2 (ore more) workers process messages from a Redis stream, which are `Event` messages from `Device` entities. 
Rules are:
- 2 workers aren't allowed to process events from the same `Device` at the same time
- Events from a given device have to be processed sequentially.

So basically, when a handler gets an `Event`, it acquires a lock (using redis store) on the associated device to prevent other events from the same device to be processed by other workers until this one has released the lock. What I did at first is to throw a `RecoverableMessageHandlingException` so that the message is retried later, but this leads to some situations where that further processing comes a little too late, when the lock has already been released a few seconds ago and other events processed (meaning sequentiality is broken).

What I'd like to do is rather that my handler tells Redis “can’t process it now, leave it to another worker (the one who acquired the lock)“, like skipping it, leaving it untouched.

Any ideas?

**Comment by MatTheCat** [comment]
Don’t know much about streams, but would it be possible to have as many streams as consumers (one consumer per stream)? If you load-balance the messages yourself I think you’ll have all the guarantees you want :thinking:

**Reply by bpolaszek** [reply]
Unfortunately,
A) AFAIK Symfony Messenger has no option to choose / change consumer / group once the worker has started
B) In my use case, device is not predictable (I won't know in advance what device will emit events, and the device list constantly grows in database)

🙁

**Reply by MatTheCat** [reply]
> A) AFAIK Symfony Messenger has no option to choose / change consumer / group once the worker has started

Yes I guess you’d need a worker per consumer.

> B) In my use case, device is not predictable

They don’t need to: eg. you can seed PHP’s RNG with a device ID and use `mt_rand` to get the associated worker index.
